Added "auto-doctor" setting to automatically apply papis doctor to new documents

f0rki opened the pull request:
Introduced a new option auto-doctor wich controls whether `papis doctor` is run on each `add`'ed|`update`d document. The idea is to automatically ensure that documents are consistent to what we expect.
(However, this flag is currently disabled by default, because it breaks a lot of tests by being on by default.)

Introduced two new config options:
* `auto-doctor = False` - whether to run doctor on all documents always
* `auto-doctor-edit = False` - whether to launch the editor on documents that cannot be fixed automatically.
* `papis (add|update) (--auto-doctor|--no-auto-doctor)` to enable/disable
  auto-doctor on the fly.

Extended the `key-exists` check with an "Autofiller" mechanism that automatically fills in missing keys. Implemented autofillers for filling `author` and `author_list` if the other one is available.

The `author_list` autofiller fixes my original issue, that I always want an `author_list` in all of my documents, because I use the `author_list` as part of the file formatting template.

Additionally, this PR also introduces auto-guessing of author separators. This allows you to pass, e.g., `--set author "John Doe, Jane Doe"` on the command line and papis will guess "," as the correct separator for the authors. Combined with the auto-doctor features, you can now do something like `papis add --set author "John Doe, Jane Doe" --set year 2023 http://example.com/paper.pdf` and 1. it will automatically fill in the `author_list`, 2. it will find that the `title` is missing and open the editor for you to fix that.

Originally, this PR stated:

> When adding a "raw" pdf, I set the `authors` key on the command line with `--set`, but also would like to have the `authors_list` in the Document. To get that I added a function that sanitizes a document once it is loaded, e.g., if `authors` is presented, but `authors_list` is not -> attempt to add it. I also added code to make sure that the `year` key is always pointing to an integer.
> 
> What do you think about having such a function?

alexfikl commented:
I definitely like the behavior since this happened to me quite a few times as well :cry: 

What do you think about adding this functionality to `papis doctor` instead? What I'm thinking of is to improve the `keys-exist` check here
https://github.com/papis/papis/blob/7677f252a4b8c1018e50e7a0e8912c54cab3d47d/papis/commands/doctor.py#L208-L225

Basically:
* Add a `fix_action` that checks if `author` exists and `author_list` does not and adds it like you did here. It should leave other keys alone and just return `None`.
* Add `author_list` to the default keys for `doctor-keys-exist-keys` in `defaults.py`, since this is very useful!

We should probably refactor `papis doctor` a bit so that we can use any checks with an autofixer in `papis add` or other places as well, but that's a bit more work.

alexfikl commented:
By the way, not sure if anybody actually uses the `papis doctor` command since it's kind of new and undocumented, but there's some barebone docs now! 
https://papis.readthedocs.io/en/latest/commands.html#module-papis.commands.doctor

jghauser commented:
I just wanted to quickly chime in to say I really appreciate this addition. Thanks! :) I've struggled with this a few times myself. I think this would fit well with `papis doctor` though I'd definitely appreciate an autofixer with `papis add` (as I tend to forget most things that aren't automatic :facepalm:).

f0rki commented:
To be honest, I was not aware of papis doctor. Seems reasonable to add this to doctor. However for my usecase I need this in `add`. I use `authors_list` in my config to rename the directory of documents, so I really need a way to set this automatically while in the add command.

How about adding a `--auto-doctor` cli switch and also a `auto-doctor = true` config to automatically run doctor on newly added documents etc.?

alexfikl commented:
> To be honest, I was not aware of papis doctor. Seems reasonable to add this to doctor. However for my usecase I need this in `add`. I use `authors_list` in my config to rename the directory of documents, so I really need a way to set this automatically while in the add command.

Completely agree! My suggestion would be just that in this PR we could just add the functionality to `papis doctor` and then do a small refactor and call it in `papis add` in a separate PR.

> How about adding a `--auto-doctor` cli switch and also a `auto-doctor = true` config to automatically run doctor on newly added documents etc.?

I'm almost inclined to always do the `auto-doctor` fixes in `papis add`. Do you think it's worth adding a switch for it?

f0rki commented:
> I'm almost inclined to always do the auto-doctor fixes in papis add. Do you think it's worth adding a switch for it?

Iguess, better to make that optional to allow users to opt-out at least using a config flag.

alexfikl commented:
> I guess, better to make that optional to allow users to opt-out at least using a config flag.

Fair enough! Sounds like a good plan to me :grin:

f0rki commented:
I updated the PR to include the auto-doctor feature. I disabled it by default, because it breaks *a lot* of tests. 

Also I have a problem with mypy when implementing the `update` method on `Document`:
```
papis/document.py:348:5: error: Signature of "update" incompatible with supertype "MutableMapping"  [override]
        def update(self, data: DocumentLike) -> None:
        ^
papis/document.py:348:5: note:      Superclass:
papis/document.py:348:5: note:          @overload
papis/document.py:348:5: note:          def update(self, SupportsKeysAndGetItem[str, Any], /, **kwargs: Any) -> None
papis/document.py:348:5: note:          @overload
papis/document.py:348:5: note:          def update(self, Iterable[tuple[str, Any]], /, **kwargs: Any) -> None
papis/document.py:348:5: note:          @overload
papis/document.py:348:5: note:          def update(self, **kwargs: Any) -> None
papis/document.py:348:5: note:      Subclass:
papis/document.py:348:5: note:          def update(self, data: Document | dict[str, Any]) -> None
Found 1 error in 1 file (checked 110 source files)
```
Any ideas on how to properly overload the `update` method, s.t., mypy doesn't complain?

alexfikl commented:
I won't have time to look at this more carefully until the weekend, but from a quick skim, I have some questions:
* Why not add the `author_list` check to the existing `keys-exist` check? Doesn't seem like a new check is necessary here.
* You're doing some custom author parsing in there. Why not use `split_author_name` from https://github.com/papis/papis/blob/main/papis/document.py#L172?
* I'm not quite sure about overloading `Document.update`. Why not just leave it as a separate function that needs to be called explicitly?

f0rki commented:
1. I will have a look on the weekend whether it makes more sense to add it to the existing check/fix.

2. I use the `split_author_names`, but was under the impression that I need to first find the right separator (i.e., to handle "Author A, Author B", "Author A and Author B", "Author A, and Author B", etc.)

3. Overloading `Document.update` ensures that wherever the document is modified it is auto-doctored. Seemed the easiest way to do that. Of course we can also call that explicitly. Maybe add that to `papis.api.save_doc`?

alexfikl reviewed (commented):
I looked over this and left some more comments, since it's a fairly big change now. Let me know if you need any help with any parts!

> * I changed it to a more generic keys-autofill check that can be extended to include more autofillers (Any suggestions here?). Extending the keys-exist seemed like it might be too restrictive for users who only want the keys-exist check.

I still think `keys-exist` is the right place for this. At least for the command-line usage, you have to explicitly pass `--fix` to apply any autofixers, so it's completely under the user's control, no?

> * I use the split_author_names, but was under the impression that I need to first find the right separator (i.e., to handle "Author A, Author B", "Author A and Author B", "Author A, and Author B", etc.). Is this not the way to go?

Ah, good catch! Yeah, I missed that in the quick skim :grin: 
I added a suggestion there to refactor that logic back into mostly `split_author_names`, but otherwise that's correct.

> * Overloading Document.update ensures that wherever the document is modified it is auto-doctored. Seemed the easiest way to do that. Of course we can also call Document.auto_doctor explicitly, whenever a new document is constructed or updated or maybe just call auto_doctor in papis.api.save_doc?

Hm, I'm not quite sure what the right way would be here. Maybe just keep the `auto_doctor` method you added and not overwrite `update`? It seems unintuitive to me that the standard `dict.update` will do work behind my back modifying the keys and values..

> @alexfikl do you want me to split the PR into the (1) author/author_list stuff for doctor and (2) the auto-doctor code?

No, I think this is fine! The `auto-doctor` part is just a small chunk apparently, so no worries.

f0rki commented on papis/commands/doctor.py:
```diff
@@ -123,7 +123,7 @@ class Error(NamedTuple):
     suggestion_cmd: str
     #: A callable that can autofix the error.
-    fix_action: FixFn
+    fix_action: Optional[FixFn]
```
This also signals whether there is a `fix_action` avilable, i.e., `fix_action is None` implies there is no `fix_action`. With `lambda: None` there is no such signal. This allows you to e.g., print a warning if no fix action is available. I use that in auto-doctor to decide whether to launch `edit` automatically. If every error returned by the checkers is fixed with `fix_action` no need to `edit` - otherwise launch the editor.

alexfikl replied:
Ah, right, missed that part of the logic! Thanks for explaining!

f0rki commented on papis/commands/doctor.py:
```diff
@@ -375,37 +375,73 @@ def key_type_check(doc: papis.document.Document) -> List[Error]:
-                                 fix_action=lambda: None,
-                                 payload=key,
-                                 doc=doc))
+            results.append(
+                Error(
+                    name=KEY_TYPE_CHECK_NAME,
+                    path=folder,
+                    msg=(
+                        "Key '{}' ({}) should be of type '{}'"
+                        " but found '{}'".format(
+                            key, doc_value, cls, type(doc_value).__name__
+                        )
+                    ),
+                    suggestion_cmd=("papis edit --doc-folder {}".format(folder)),
+                    fix_action=make_fixer(key, cls),
+                    payload=key,
+                    doc=doc,
+                )
+            )
```
oops. sorry. yeah I can do that.

f0rki commented on papis/commands/doctor.py:
```diff
@@ -375,37 +375,73 @@ def key_type_check(doc: papis.document.Document) -> List[Error]:
         except Exception as exc:
-            logger.error("Invalid (key, type) pair: '%s'.",
-                         value, exc_info=exc)
+            logger.error("Invalid (key, type) pair: '%s'.", value, exc_info=exc)
```
yes. I usually use an auto-formatter. I tried to not commit those changes but maybe I accidentally `git add`ed them.

f0rki commented on papis/commands/doctor.py:
```diff
@@ -508,8 +544,186 @@ def fixer() -> None:
+            authors = []
+            if "author" in doc:
+                authors = doc["author"]
+            if "authors" in doc:
```
This is mostly me forgetting, whether it is `author` or `authors`. I can remove that.

f0rki commented:
@alexfikl I think I addressed most of your comments, except for this one:

> This whole thing looks over-engineered to me, I'm afraid. Other checks have "autofillers" already (e.g. for refs) and they're a lot simpler.

:shrug: maybe, but seems to me the most sensible/extensible way to handle this. If this blocks merging I can change it to something else...

alexfikl reviewed (commented):
Left a few more nitpicks here and there, but otherwise the code parts are looking a lot better to me! I'm definitely fine with leaving the `Autofiller` stuff in; it was just so different to everything else that it stood out to me during review, haha. If it works, it works!

I'll try to use this branch a bit a see how it works. @jghauser @alejandrogallo Can you give this a look as well?

f0rki commented on papis/commands/doctor.py:
```diff
@@ -534,6 +695,38 @@ def run(doc: papis.document.Document, checks: List[str]) -> List[Error]:
+    import papis.commands.edit
+    from papis.api import save_doc
+
+    with auto_doctor_recursion_guard() as in_auto_doctor:
```
yes. Actually, with the previous way of doing it in `Document.update` the same could happen. Just lucky it didn't, but any code called by `papis doctor` could've triggered (endless) recursion by called `Document.update` instead of using the `[]` access.

f0rki commented on tests/test_document.py:
```diff
@@ -162,3 +162,36 @@ def test_dump() -> None:
     assert result == expected_result
+
+
+def test_author_separator_heuristics() -> None:
+    import re
+    from papis.document import determine_authors_separator
+
+    def is_comma_and_re(sep):
+        assert sep
+        assert re.match(sep, ", and")
+        assert re.match(sep, ",and")
+        assert re.match(sep, ",")
+
+    # TODO: properly check for family/given splits
+
+    s = "Sanger, F. and Nicklen, S. and Coulson, A. R."
+    assert determine_authors_separator(s) == "and"
+    s = "Fabian Sanger and Steven Nicklen and Alexander R. Coulson"
+    assert determine_authors_separator(s) == "and"
+    s = "Fabian Sanger, Steven Nicklen, Alexander R. Coulson"
+    assert determine_authors_separator(s) == ","
+    s = "Fabian Sanger, and Steven Nicklen, and Alexander R. Coulson"
+    sep = determine_authors_separator(s)
+    is_comma_and_re(sep)
+    s = "Fabian Sanger, Steven Nicklen, and Alexander R. Coulson"
+    sep = determine_authors_separator(s)
+    is_comma_and_re(sep)
+    s = "John Doe, Jane Dorian, and James T. Unknown "
+    sep = determine_authors_separator(s)
+    is_comma_and_re(sep)
+    s = "Dagobert Duck and von Beethoven, Ludwig and Ford, Jr., Henry"
+    assert determine_authors_separator(s) == "and"
+    s = "Turing, A. M."
+    assert determine_authors_separator(s) is None
```
I think I wanted to expand this to use the split function and actually check the returned data (i.e., `assert r[0]['family'] == 'Sanger'`).

f0rki commented on papis/document.py:
```diff
@@ -181,13 +237,31 @@ def split_authors_name(authors: List[str],
+        else:
+            subauthor_iter = re.split(r"\s*{}\s+".format(sep), subauthors)
+
+        if not subauthor_iter:
+            continue
```
ah yes. I was getting impatient, because I got `None` somewhere in that code. will update.

f0rki commented on papis/document.py:
```diff
@@ -181,13 +237,31 @@ def split_authors_name(authors: List[str],
+        if not subauthors:
+            continue
+        sep = None
+        if separator is None:
+            sep = determine_authors_separator(subauthors)
+        if sep is None:
+            subauthor_iter = [subauthors]
+        else:
+            subauthor_iter = re.split(r"\s*{}\s+".format(sep), subauthors)
```
It returns `None` if it determines a single author to handle the `Turing, A. M.` case.

alexfikl replied:
Yeah, this was mostly a note if you change it to return "and" in that case as well.

f0rki commented on papis/document.py:
```diff
@@ -169,8 +169,64 @@ def author_list_to_author(data: Dict[str, Any]) -> str:
+        sep = r",\s*(?:and)?"
+    elif _AUTHOR_SINGLE_AUTHOR_FAMILY_FIRST_RE.match(authors):
+        # found a single author "Last, First"
+        sep = None
```
I thought one less `re.split` is probably better, but yeah, probably works too.

alexfikl replied:
Yeah, that's a fair point, so up to you! I was mostly thinking of removing a bunch of `if sep is None` in the calling code.

f0rki commented:
ok fixed the remaining tests. CI is green now.

f0rki commented:
I just noticed a problem: it seems temporary directories leak into the papis database cache.

alexfikl commented:
Sorry about letting this one fall through the cracks as well :( We definitely still want to get this in, but it's getting quite large with changes all over the place.

Just to re-iterate a bit what the state of this is (in my mind at least)
1. There seems to be a lot of trouble with auto-saving the doctor changes / infinite recursion / duplicate documents. This makes me believe that there's some architecture issue here and we should fix that first instead of hacking around it.
2. The changes to `key_type_check` can go in separately.
3. The auto-filler parts can also go in separately.
4. `get_documents() might actually change the current document on disk, so we need to reload here` This sounds like a big bug. Is this related to the auto-doctor stuff or is it present on `main`? 
5. The additions to `split_authors_name` are done and can go in separately as well.

What do you think about making separate PRs for 2, 3 and 4 so we can get those in faster?

For 1, it seems like this all stems from doing `auto_doctor` on save. Is that accurate or do you think something else is complicating the implementation?

f0rki commented:
> What do you think about making separate PRs for 2, 3 and 4 so we can get those in faster?

Makes sense. I'll see when I get to it, but my time is somewhat limited at the moment.

> get_documents() might actually change the current document on disk, so we need to reload here This sounds like a big bug. Is this related to the auto-doctor stuff or is it present on main?

I believe it is also present on main, although it might not be triggered anywhere except for the auto-doctor code.

> There seems to be a lot of trouble with auto-saving the doctor changes / infinite recursion / duplicate documents. This makes me believe that there's some architecture issue here and we should fix that first instead of hacking around it.

* The infinite recursion guard is probably not needed anymore, since auto-doctor is now applied in the command code (i.e., `commands/update.py` and   `commands/add.py`) and not in the document class itself.
* I only encountered the duplicate documents issue in the papis backend, so that seems to be an issue there.

alexfikl commented:
> Makes sense. I'll see when I get to it, but my time is somewhat limited at the moment.

That's quite ok. Happy to help a bit with these!

alexfikl commented:
@f0rki Went ahead and rebased this and fixed the many many conflicts :cry:

I cleaned up some other stuff (like that recursive guard thing), so this is looking pretty nice at the moment. Some simple tests also seems to work nicely! Some more things that would be nice to get done:

* Are we ok with the `--auto-doctor` flag? Maybe `--auto-fix` or something? I'm not sure what would be more intuitive here..
* That bug with `get_documents` and reloading, mentioned in https://github.com/papis/papis/pull/598#issuecomment-1653158431, would be nice to figure out in more detail. I don't think that's an issue with the current version of the code, since it doesn't try to add documents to the database repeatedly anywhere, but it should yell if we try to do something stupid :\\
* Need some tests for the auto-doctor functionality. Should just find an DOI somewhere that needs some small fixes.

@f0rki Hopefully I kept all the functionality you needed, but you'll probably need to look over it again :worried: 
@jghauser If you have a chance to test this on some actual documents that need auto-fixing would be great!

alexfikl commented:
I think we've been sitting on this enough at this point, so I'll go ahead and merge it :rocket: 

@f0rki Thanks a lot for working on this! (and sorry for the delays)
